libcodechecker/server/client_db_access_handler.py

```diff
@@ -1546,50 +1546,45 @@ def replaceConfigInfo(self, run_id, config_values):
             return False
 
     @timeit
-    def needFileContent(self, run_id, filepath):
+    def needFileContent(self, filepath, content_hash):
         """
         """
-        try:
-            f = self.__session.query(File) \
-                .filter(and_(File.run_id == run_id,
-                             File.filepath == filepath)) \
-                .one_or_none()
-        except Exception as ex:
-            raise shared.ttypes.RequestFailed(
-                shared.ttypes.ErrorCode.GENERAL,
-                str(ex))
-
-        run_inc_count = self.__session.query(Run).get(run_id).inc_count
-        needed = False
+        f = self.__session.query(File) \
+            .filter(and_(File.content_hash == content_hash,
+                         File.filepath == filepath)) \
+            .one_or_none()
         if not f:
-            needed = True
-            f = File(run_id, filepath)
-            self.__session.add(f)
-            self.__session.commit()
-        elif f.inc_count < run_inc_count:
-            needed = True
-            f.inc_count = run_inc_count
-            self.__session.commit()
+            try:
+                self.__session.commit()
+                f = File(filepath, content_hash)
+                self.__session.add(f)
+                self.__session.commit()
+            except sqlalchemy.exc.IntegrityError:
+                # Other transaction might added the same file in the meantime.
+                self.__session.rollback()
+
+        needed = self.__session.query(FileContent).get(content_hash) is None
         return NeedFileResult(needed, f.id)
 
     @timeit
-    def addFileContent(self, fid, content, encoding):
+    def addFileContent(self, content_hash, content, encoding):
```

Comment:
I think it is redundant. We already trust the client in other places, and hashing is not cheap.